towerlib/towerlib.py

```diff
@@ -1997,24 +2034,707 @@ def schedules(self):
+
+        """
+        credential_types = []
+        for item in self.credential_types:
+            credential_types.append(item)
+        return credential_types
+
+    def get_credential_id_from_existing_project(self, scm_url):
+        """This function gets credential id from an existing project, which has the same credential id.
+
+        This function is for additional convenience. If there is already a project with the same scm url configured,
+        then we will find it with this function. This function also makes sure that the user does not have to specify a
+        credential id in the config file .
+        This function will only be called if no credential id is given in the config file. Anyway, if no credential is
+        given by config file and no credential id is found with this function, the script will continue and create
+        projects, because projects can be created in Ansible Tower without credentials (but in that case, the scm
+        update job will fail.)
+
+        Args:
+            scm_url: the scm_url of the corresponding project.
+
+        Returns:
+            id of a credential if corresponding project exists, else None.
+
+        """
+        scm_credential_type_ids = []
+        credential_types = self.get_all_credential_types()
+        for credential_type in credential_types:
+            if credential_type.kind == 'scm':
+                scm_credential_type_ids.append(credential_type.id)
+        credential_id = None
+        all_projects = self.get_all_projects()
+        for scm_credential_type_id in scm_credential_type_ids:
+            for project in all_projects:
+                if project.scm_url == scm_url:
+                    self._logger.debug("Project found with the scm credential.")
+                    if project.credential.credential_type.id == scm_credential_type_id:
+                        credential_id = project.credential.id
+                        if credential_id:
+                            self._logger.debug("Credential ID found from the existing project '{}' and the id is '{}'"
+                                               .format(project.name, credential_id))
+                            return credential_id
+        return credential_id
+
+    def get_project_id_by_branch_name(self, scm_url, scm_branch):
+        """Get the ids of the ansible tower project with given scm_url and branch name.
+
+        Args:
+            scm_url: given scm url.
+            scm_branch: given scm branch.
+
+        Returns:
+            list: list of project ids
+
+        """
+        project_ids = []
+        projects = self.get_all_projects()
+        for project in projects:
+            if project.scm_type == 'git':
+                if project.scm_url == scm_url and project.scm_branch == scm_branch:
+                    project_ids.append(project.id)
+        return project_ids
+
+    def get_production_branch_project_id(self, scm_url):
```

Comment:
I think that this functionality holds too much client side business logic to be generic and useful to people using the library. It is not standard to have 'production' branch names. A more reusable approach would be to have a method like ```get_projects_ids_by_branch_name```.